js/classes/entities/midground/transport_portal.js

```diff
@@ -0,0 +1,133 @@
+class TransportPortal {
+    /**
+     * @param {Vector} blockPos The position of the block the portal is on
+     * @param {Number} type The type of portal (e.g. TransportPortal.PURPLE or TransportPortal.YELLOW)
+     * @author Nathan Hinthorne
+     */
+    constructor(blockPos, type) {
+        this.pos = Vector.blockToWorldSpace(blockPos);
+        this.type = type;
+
+        /** The number of enemies to release at once. */
+        this.spawnGroupSize = TransportPortal.SPAWN_GROUP_SIZE;
+
+        this.animations = [];
+        this.loadAnimations();
+
+        /** Whether the portal is active, in the middle of activating, or inactive */
+        this.action = "inactive";
+
+        switch (this.type) {
+            case TransportPortal.PURPLE:
+                this.particleType = ParticleEffect.AURA_PURPLE;
+                this.borderColor = COLORS.PURPLE;
+                break;
+            case TransportPortal.YELLOW:
+                this.particleType = ParticleEffect.AURA_YELLOW;
+                this.borderColor = COLORS.YELLOW;
+                break;
+        }
+
+        this.activationTimer = TransportPortal.ACTIVATION_DELAY;
+        this.particleTimer = 0;
+        this.center = Vector.add(this.pos, new Vector(TransportPortal.SCALED_SIZE.x / 2, TransportPortal.SCALED_SIZE.y / 2));
+    };
+
+    activate() {
+        this.action = "activating";
+        ASSET_MGR.playSFX(SFX.PORTAL_ACTIVATE.path, SFX.PORTAL_ACTIVATE.volume);
+
+        // hardcode this portal's border to bring you to the endzone
+        this.border = new Border(
+            Vector.add(this.pos, new Vector(16, 23)),
+            Vector.blockToWorldSpace(new Vector(1, 3)),
+            Zone.getZones().end.endZone
+        );
+
+        GAME.addEntity(this.border);
+    }
+
+    update() {
+        // if in the middle of activating, count down to activation
+        if (this.action === "activating") {
+            this.activationTimer -= GAME.clockTick;
+            if (this.activationTimer <= 0) {
+                this.action = "active";
+                ASSET_MGR.playSFX(SFX.PORTAL_IDLE.path, SFX.PORTAL_IDLE.volume);
+            }
+        }
+
+        // make portal perform a circular motion
+        this.pos.x += Math.cos(GAME.gameTime * 2) / 10;
+        this.pos.y += Math.sin(GAME.gameTime * 2) / 10;
+
+        if (this.action != "active") return;
```

Comment:
Should this happen sooner? Not sure when it makes sense for this to happen, just seems like it should be one of the first checks.

Reply:
No, I think we should check if the portal is active later. I want the portal to perform its circular motion even when its not active